dora/core/server/master/src/main/java/alluxio/master/scheduler/Scheduler.java

```diff
@@ -139,41 +136,32 @@ public boolean enqueueTaskForWorker(@Nullable WorkerInfo workerInfo, Task task,
+        // track running tasks of a job
+        ConcurrentHashSet<Task<?>> tasks = mJobToRunningTasks.computeIfAbsent(task.getJob(),
                 j -> new ConcurrentHashSet<>());
-            tasks.add(task);
-            task.getResponseFuture().addListener(() -> {
-              Job job = task.getJob();
-              try {
-                job.processResponse(task); // retry onfailure logic inside
-              /* TODO(lucy) now whether task succeed or fail, remove it from q,
-              it could be add numOfRetry logic to the task with added upgrade/degrade
-              priority to add back to the q for retry. */
-                workerTaskQ.remove(task);
-                ConcurrentHashSet runningTaskSet = mJobToRunningTasks.getOrDefault(
-                    job, new ConcurrentHashSet<>());
-                runningTaskSet.remove(task);
-                // TODO(lucy) currently processJob is only called in the single
-                // threaded scheduler thread context, in future once tasks are
-                // completed, they should be able to call processJob to resume
-                // their own job to schedule next set of tasks to run.
-              } catch (Exception e) {
-                // Unknown exception. This should not happen, but if it happens we don't
-                // want to lose the worker thread, thus catching it here. Any exception
-                // surfaced here should be properly handled.
-                LOG.error("Unexpected exception thrown in response future listener.", e);
-                job.failJob(new InternalRuntimeException(e));
-              }
-            }, mScheduler.getWorkingExecutor());
+        tasks.add(task);
+        task.execute(blkWorkerClientResource.get(), workerInfo);
+        task.getResponseFuture().addListener(() -> {
+          Job job = task.getJob();
+          try {
+            job.processResponse(task); // retry on failure logic inside
+            // TODO(lucy) currently processJob is only called in the single
+            // threaded scheduler thread context, in future once tasks are
+            // completed, they should be able to call processJob to resume
+            // their own job to schedule next set of tasks to run.
           } catch (Exception e) {
-            LOG.error("Unexpected exception thrown in submitting task:{} of job:{}.",
-                task, task.getJob(), e);
+            // Unknown exception. This should not happen, but if it happens we don't
+            // want to lose the worker thread, thus catching it here. Any exception
+            // surfaced here should be properly handled.
+            LOG.error("Unexpected exception thrown in response future listener.", e);
+            job.failJob(new InternalRuntimeException(e));
+          } finally {
+            // whether task succeed or fail, remove it from q,
+            workerTaskQ.remove(task);
+            ConcurrentHashSet<Task<?>> runningTaskSet = mJobToRunningTasks.get(job);
```

Comment:
change to get since it should have the key in the task callback, otherwise it's a bug